pubsub/google/cloud/pubsub_v1/publisher/batch/mp.py

```diff
@@ -0,0 +1,398 @@
+            not text. You must send data as a bytestring
+            (``bytes`` in Python 3; ``str`` in Python 2), and this library
+            will raise an exception if you send a text string.
+
+            The reason that this is so important (and why we do not try to
+            coerce for you) is because Pub/Sub is also platform independent
+            and there is no way to know how to decode messages properly on
+            the other side; therefore, encoding and decoding is a required
+            exercise for the developer.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        Args:
+            data (bytes): A bytestring representing the message body. This
+                must be a bytestring (a text string will raise TypeError).
+            attrs (Mapping[str, str]): A dictionary of attributes to be
+                sent as metadata. (These may be text strings or byte strings.)
+
+        Raises:
+            TypeError: If the ``data`` sent is not a bytestring, or if the
+                ``attrs`` are not either a ``str`` or ``bytes``.
+
+        Returns:
+            ~.pubsub_v1.publisher.batch.mp.Future: An object conforming to the
+                :class:`concurrent.futures.Future` interface.
+        """
+        # Sanity check: Is the data being sent as a bytestring?
+        # If it is literally anything else, complain loudly about it.
+        if not isinstance(data, six.binary_type):
+            raise TypeError('Data being published to Pub/Sub must be sent '
+                            'as a bytestring.')
+
+        # Coerce all attributes to text strings.
+        for k, v in copy.copy(attrs).items():
+            if isinstance(data, six.text_type):
+                continue
+            if isinstance(data, six.binary_type):
+                attrs[k] = v.decode('utf-8')
+                continue
+            raise TypeError('All attributes being published to Pub/Sub must '
+                            'be sent as text strings.')
+
+        # Store the actual message in the batch's message queue.
+        self._shared.messages.append(
+            types.PubsubMessage(data=data, attributes=attrs),
+        )
+
+        # Add the size of the message to our size tracking.
+        self._size += len(data)
+        self._size += sum([len(k) + len(v) for k, v in attrs.items()])
+
+        # Return a Future. That future needs to be aware of the status
+        # of this batch.
+        f = Future(self._shared)
+        self._shared.futures.append(f)
+
+        # Check and see if we have hit message limits. If we have,
+        # commit.
+        if len(self._shared.messages) >= self.settings.max_messages:
+            self._shared.status = 'at message cap'
+            self.commit()
+        if self._size >= self.settings.max_bytes:
```

Comment:
This won't work. We want max_bytes to be a strict upper limit here, so that users cannot exceed message size limits due to our batching. For example, if a user sends a 1 byte message and a 10 MB message, and the limit is set to the max of 10 MB, the messages should be sent separately. If sent together they will fail at CPS since they are greater than 10 MB. What we've done in other languages is send the batch and make the message that was published the first message of a new batch.

You also don't handle large messages here. For example if the limit is 10 KB, and we have a batch of 5 KB and a user publishes 15 KB, we should publish them both, since we've exceeded the limit we want to publish the batch, and we want to send the large message directly since it already exceeds the limit on its own.